src/Microsoft.ML.DataView/TypeManager.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.ML.Internal.DataView;
+
+namespace Microsoft.ML.Data
+{
+    /// <summary>
+    /// A singleton class for managing the map between ML.NET <see cref="DataViewType"/> and C# <see cref="Type"/>.
+    /// To support custom column type in <see cref="IDataView"/>, the column's underlying type (e.g., a C# class's type)
+    /// should be registered with a class derived from <see cref="DataViewType"/>.
+    /// </summary>
+    public static class DataViewTypeManager
+    {
+        // Types have been used in ML.NET type systems. They can have multiple-to-one type mapping.
+        // For example, UInt32 and Key can be mapped to uint. This class enforces one-to-one mapping for all
+        // user-registered types.
+        private static HashSet<Type> _bannedRawTypes = new HashSet<Type>()
+        {
+            typeof(Boolean), typeof(SByte), typeof(Byte),
+            typeof(Int16), typeof(UInt16), typeof(Int32), typeof(UInt32),
+            typeof(Int64), typeof(UInt64), typeof(Single), typeof(Double),
+            typeof(string), typeof(ReadOnlySpan<char>)
+        };
+
+        private static Dictionary<Type, DataViewType> _rawTypeToDataViewTypeMap = new Dictionary<Type, DataViewType>();
+        private static Dictionary<DataViewType, Type> _dataViewTypeToRawTypeMap = new Dictionary<DataViewType, Type>();
+        private static SpinLock _lock = new SpinLock();
```

Comment:
Why use a SpinLock, and not just a regular `lock`?

Reply:
From .NET doc,
```
A mutual exclusion lock primitive that causes the thread that is trying to acquire the lock to wait in a loop, or spin, for a period of time before yielding its quantum. In scenarios where the wait for the lock is expected to be short, SpinLock offers better performance than other forms of locking. For more information, see SpinLock.
```
Because we only lock for a few instructions, I feel `SpinLock` can provide better performance.

---
In reply to: [285626445](https://github.com/dotnet/machinelearning/pull/3263#discussion_r285626445) [](ancestors = 285626445)

Reply:
My suggestion would be to gather data to support the `I feel` statement above. See the answer from Stephen Toub here:

https://social.msdn.microsoft.com/Forums/en-US/b5b8373a-6d56-427f-b90b-1f67f78010c4/spinlock-usageperformance

Reply:
Let me just use `lock`. This is definitely not a computation  bottleneck and `lock` can simplify my code a lot.

---
In reply to: [285669394](https://github.com/dotnet/machinelearning/pull/3263#discussion_r285669394) [](ancestors = 285669394)